### PR TITLE
Update token cache logic

### PR DESCRIPTION
**Pull request by y-myajima**
# Description

- change update token cahe logic
  - update token partially (allow some request to fail)    
  - same as [client sidecar](https://github.com/AthenZ/athenz-client-sidecar/blob/5f0eb1b1090a97303ee3a4549901152335ce7026/service/role.go#L274-L292)

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code

## Related issue/PR

**Delete this section if there are no issues or pull requests that relate to this pull request.**
- Fixes #_issue_
- Closes #_PR_

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [x] Passed all pipeline checking

## Checklist for maintainer

- Use `Squash and merge`
- Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- Delete the branch after merge

### CODE REVIEW

File: pkg/token/daemon.go [hunk 1]
```diff
@@ -143,26 +143,94 @@ func newDaemon(idConfig *config.IdentityConfig, tt mode) (*daemon, error) {
...
+		defer close(echan)
+		atErrorCount := 0
+		for _, t := range atTargets {
+			key := t // prevent closure over loop variable
+			err := d.updateTokenWithRetry(key, mACCESS_TOKEN)
+			if err != nil {
+				echan <- err
+				atErrorCount += 1
+			}
+		}
```
Comment (WindzCUHK) [inline pkg/token/daemon.go:161]:
```suggestion
		var wg sync.WaitGroup
		for _, t := range atTargets {
		    wg.Add(1)
		    go func(key string) {
                        defer wg.Done()
                        err := d.updateTokenWithRetry(key, mACCESS_TOKEN)
			if err != nil {
				echan <- err
				atErrorCount += 1 // change to atomic.Int
			}
                   }(t)
		}
```
client sidecar do parallel based on key, 1 key = 1 routine.
please use https://gobyexample.com/waitgroups to wait for all routines to finish.

File: pkg/token/daemon.go [hunk 2]
```diff
@@ -172,7 +240,8 @@ func (d *daemon) writeFiles() error {
 	}
 
 	w := util.NewWriter()
-	d.accessTokenCache.Range(func(k CacheKey, t Token) error {
+
+	err := d.accessTokenCache.Range(func(k CacheKey, t Token) error {
```
Comment (WindzCUHK) [inline pkg/token/daemon.go:244]: the current write file logic is all or nothing (NOT best effort), is it expected?

File: pkg/token/daemon.go [hunk 3]
```diff
@@ -143,26 +145,100 @@ func newDaemon(idConfig *config.IdentityConfig, tt mode) (*daemon, error) {
...
 }
 
-func (d *daemon) updateTokenWithRetry() error {
+func (d *daemon) updateTokenCaches() <-chan error {
+	var atErrorCount, rtErrorCount atomic.Int64
+	atTargets := d.accessTokenCache.Keys()
+	rtTargets := d.roleTokenCache.Keys()
+	log.Infof("Attempting to fetch tokens from Athenz ZTS server: access token targets[%v], role token targets[%v]...", atTargets, rtTargets)
+	echan := make(chan error, len(atTargets)+len(rtTargets))
+	go func() {
```
Comment (WindzCUHK) [inline pkg/token/daemon.go:154]:
```suggestion
```
no need for `go func() {`, just let the function to be blocking.

`for err := range d.updateTokenCaches() {` will always block until the channel close.

Comment (y-myajima) [reply pkg/token/daemon.go]: fixed: https://github.com/AthenZ/k8s-athenz-sia/pull/75/commits/24f9c761a3156502844efccc31dbfaa3cd52a193

File: pkg/token/daemon.go (hunk 1 above)
Comment (y-myajima) [reply pkg/token/daemon.go]: fixed: https://github.com/AthenZ/k8s-athenz-sia/pull/75/commits/471fb65a87c711df633ef710677f9990eceb26fe

**Review by WindzCUHK — APPROVED** [review]

**Pull request merged**